### PR TITLE
Use reflection to call cleaner on direct byte buffers in JDK9

### PR DESCRIPTION
**Pull request by varming**
In JDK9 `sun.misc.Cleaner` is replaced with `java.lang.ref.Cleaner$Cleanable`, a class with a method `void clean()` just like the old cleaner. By using a method reference we don't need to know if the class is one or the other.

### CODE REVIEW

File: common/src/main/java/io/netty/util/internal/Cleaner0.java [hunk 1]
```diff
@@ -61,9 +66,9 @@ static void freeDirectBuffer(ByteBuffer buffer) {
             return;
         }
         try {
-            Cleaner cleaner = (Cleaner) PlatformDependent0.getObject(buffer, CLEANER_FIELD_OFFSET);
+            Object cleaner = PlatformDependent0.getObject(buffer, CLEANER_FIELD_OFFSET);
             if (cleaner != null) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/Cleaner0.java]: @varming do we also need to check if `CLEAN_METHOD != null` ?

**Comment by normanmaurer** [comment]
@varming one question...

Could you also please update the commit message to match our template:

http://netty.io/wiki/writing-a-commit-message.html

I will also setup a CI job that will test netty with java9 today.

File: common/src/main/java/io/netty/util/internal/Cleaner0.java (hunk 1 above)
Comment (varming) [inline common/src/main/java/io/netty/util/internal/Cleaner0.java]: CLEAN_METHOD == null implies CLEANER_FIELD_OFFSET == -1, so I think we are good.

Usually I would add `assert CLEAN_METHOD != NULL : "CLEANER_FIELD_OFFSET != -1 implies CLEAN_METHOD != NULL";` before the try block, but I don't know what your policy on assert is.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/Cleaner0.java]: asserts are fine... Just add it :)

**Comment by varming** [comment]
Oops, I forgot about the commit message template. Fixed.

**Comment by normanmaurer** [comment]
No worries please fix for the other prs as well

**Comment by varming** [comment]
I updated the other PRs.

BTW. It looks like the test run got stuck.

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (34d8b514d7a3b62e6101bade10a445bc448a0e65) and 4.0 (5b92adae1d5529010410d64644e6f039779fd718)

**Pull request closed without merging**